server/src/main/java/com/thoughtworks/go/server/persistence/AgentDao.java

```diff
@@ -78,35 +104,271 @@ public Agent agentByUuid(String uuid) {
     public void associateCookie(final AgentIdentifier agentIdentifier, final String cookie) {
         final String uuid = agentIdentifier.getUuid();
         final String key = agentCacheKey(uuid);
-        synchronized (key) {
+
+        List<String> uuids = singletonList(agentIdentifier.getUuid());
+        AgentMutex mutex = agentMutexes.acquire(uuids);
+        synchronized (mutex) {
             transactionTemplate.execute(new TransactionCallbackWithoutResult() {
-                @Override protected void doInTransactionWithoutResult(TransactionStatus status) {
-                    Agent agent = fetchAgentByUuid(uuid);
+                @Override
+                protected void doInTransactionWithoutResult(TransactionStatus status) {
+                    Agent agent = fetchAgentFromDBByUUID(uuid);
                     if (agent == null) {
-                        agent = new Agent(uuid, cookie, agentIdentifier.getHostName(), agentIdentifier.getIpAddress());
+                        agent = new Agent(uuid, agentIdentifier.getHostName(), agentIdentifier.getIpAddress(), cookie);
```

Comment:
Wouldn't this be a security issue? Now any unregistered agent can make a request to associate cookie and can end up as registered and enabled in GoCD.

Reply:
Agreed. Now, `AgentDao.associateCookie` will throw an `UnregisteredAgentException` if the said agent is not present in the DB.